### PR TITLE
Enable authentication for the HTTP interface.

### PR DESCRIPTION
**Pull request by royjacobson**
CC @Pothulapati @ashotland for awareness.

### CODE REVIEW

File: src/server/main_service.cc [hunk 1]
```diff
@@ -2029,6 +2030,11 @@ GlobalState Service::GetGlobalState() const {
 }
 
 void Service::ConfigureHttpHandlers(util::HttpListenerBase* base) {
+  // We set the password for the HTTP service unless it is only enabled on the
+  // admin port and the admin port is password-less.
+  if (GetFlag(FLAGS_primary_port_http_enabled) || !GetFlag(FLAGS_admin_nopass)) {
```
Comment (chakaz) [inline src/server/main_service.cc:2035]: Why this logic?
Why not use `--requirepass` for non-admin HTTP connections?
Is the admin-port HTTP page different than the non-admin one?

**Review by dranikpg — APPROVED** [review]

File: src/server/main_service.cc (hunk 1 above)
Comment (royjacobson) [reply src/server/main_service.cc:2035]: `GetPassword()` checks `requirepass` internally so it's used for both admin and non-admin connections. They're the same HTTP page.

**Comment by ashotland** [comment]
Hi @royjacobson, what should we watch for?

Does http on admin approt require no auth? (we hit it for metrics)

File: src/server/main_service.cc (hunk 1 above)
Comment (chakaz) [reply src/server/main_service.cc:2035]: Gotcha

**Review by chakaz — APPROVED** [review]

**Comment by royjacobson** [comment]
> Hi @royjacobson, what should we watch for?
> 
> Does http on admin approt require no auth? (we hit it for metrics)

If the admin interface is configured without password and the HTTP interface is not exposed on the main port (`--noprimary_port_http_enabled`) then there's no change.

**Comment by ashotland** [comment]
> > Hi @royjacobson, what should we watch for?
> > Does http on admin approt require no auth? (we hit it for metrics)
> 
> If the admin interface is configured without password and the HTTP interface is not exposed on the main port (`--noprimary_port_http_enabled`) then there's no change.

Just for completeness, yes this is the case, we are running as follows:
--primary_port_http_enabled=false
--admin_port=9999
--admin_nopass=true

**Comment by romange** [comment]
@royjacobson  can you please provide an example on how to fetch metrics with password using curl/wget?

**Comment by s5unty** [comment]
In versions after 1.9, under the default configuration, the `/metrics` endpoint requires basic authentication for access.
Due to the limitations of Prometheus's `scrape_config`, each job can only specify a single basic authentication setting.
To collect from different authenticated `/metrics` endpoints within a single job, you can set these three configurations.

```
--primary_port_http_enabled=false
--admin_nopass=true
--admin_port=16579
```
This way, the /metrics endpoint will be provided on the port `16579` and will not require basic authentication.

**Comment by romange** [comment]
@royjacobson  I prefer cancelling http authentication and rethink it properly as a separate task. Too many things are colliding around it.

**Comment by romange** [comment]
@s5unty  can you explain in more details why setting `basic_auth` in scrape_config does not work for you?
https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config

**Comment by s5unty** [comment]
> @s5unty can you explain in more detail why setting `basic_auth` in scrape_config does not work for you? https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config

To scrape the `/metrics` endpoints of two instances while keeping the `job_name` uniform,

- case1: `found multiple scrape configs with job name "my-job"`

    ```
    - job_name: 'my-job'
      metrics_path: /metrics
      static_configs:
        - targets: [ 10.1.1.1:6479 ]
      basic_auth: { username: "user", password: "pw6479" }
    
    - job_name: 'my-job'
      metrics_path: /metrics
      static_configs:
        - targets: [ 10.1.1.1:6579 ]
      basic_auth: { username: "user", password: "pw6579" }
    ```

- case2: `static_configs` does not support `basic_auth`

    ```
    - job_name: 'my-job'
      metrics_path: /metrics
      static_configs:
        - targets: [ 10.1.1.1:6479 ]
          basic_auth: { username: "user", password: "pw6479" }
    
        - targets: [ 10.1.1.1:6579 ]
          basic_auth: { username: "user", password: "pw6579" }
    ```

**Pull request merged**